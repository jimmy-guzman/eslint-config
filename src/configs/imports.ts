import { createTypeScriptImportResolver } from "eslint-import-resolver-typescript";
import importX from "eslint-plugin-import-x";
import nodePlugin from "eslint-plugin-n";

import type { TypedConfigItem } from "../types";

import { importsRules } from "../rules/imports";

const importsTypescriptConfig = () => {
  const { rules, settings } = importX.configs.typescript;

  return [
    {
      name: "jimmy.codes/imports/typescript",
      rules,
      settings: {
        "import-x/extensions": settings["import-x/extensions"],
        "import-x/external-module-folders":
          settings["import-x/external-module-folders"],
        "import-x/parsers": settings["import-x/parsers"],
        "import-x/resolver-next": [
          createTypeScriptImportResolver({
            alwaysTryTypes: true,
          }),
        ],
      },
    },
  ];
};

interface ImportsConfigOptions {
  typescript?: boolean;
}

export const importsConfig = ({
  typescript = false,
}: ImportsConfigOptions = {}) => {
  return [
    {
      name: "jimmy.codes/imports",
      plugins: {
        "import-x": importX,
        "n": nodePlugin,
      },
      rules: importsRules,
    },
    ...(typescript ? importsTypescriptConfig() : []),
  ] as const satisfies TypedConfigItem[];
};
