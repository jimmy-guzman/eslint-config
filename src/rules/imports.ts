import importX from "eslint-plugin-import-x";

import type { Rules } from "../types";

export const importsRules = {
  ...importX.configs.recommended.rules,
  "import-x/consistent-type-specifier-style": ["error", "prefer-top-level"],
  "import-x/first": "error",
  // ! can't get this rule to work
  "import-x/namespace": "off",
  "import-x/newline-after-import": "error",
  "import-x/no-absolute-path": "error",
  "import-x/no-empty-named-blocks": "error",
  "import-x/no-self-import": "error",
  "import-x/no-useless-path-segments": "error",
} satisfies Rules;
