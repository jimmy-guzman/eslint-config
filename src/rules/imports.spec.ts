import { importsRules } from "./imports";

test("should create imports rules", () => {
  expect(importsRules).toMatchSnapshot();
});
