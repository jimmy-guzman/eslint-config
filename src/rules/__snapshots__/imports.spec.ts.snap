// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should create imports rules 1`] = `
{
  "import-x/consistent-type-specifier-style": [
    "error",
    "prefer-top-level",
  ],
  "import-x/default": "error",
  "import-x/export": "error",
  "import-x/first": "error",
  "import-x/named": "error",
  "import-x/namespace": "off",
  "import-x/newline-after-import": "error",
  "import-x/no-absolute-path": "error",
  "import-x/no-duplicates": "warn",
  "import-x/no-empty-named-blocks": "error",
  "import-x/no-named-as-default": "warn",
  "import-x/no-named-as-default-member": "warn",
  "import-x/no-self-import": "error",
  "import-x/no-unresolved": "error",
  "import-x/no-useless-path-segments": "error",
  "node-import/prefer-node-protocol": "error",
}
`;
